src/Npgsql/Internal/TypeInfoMapping.cs

```diff
@@ -0,0 +1,205 @@
+using System;
+using System.Collections.Generic;
+using Npgsql.Internal.Converters;
+using Npgsql.PostgresTypes;
+
+namespace Npgsql.Internal;
+
+/// <summary>
+/// 
+/// </summary>
+/// <param name="options"></param>
+/// <param name="mapping"></param>
+/// <param name="resolvedDataTypeName">Signals whether a resolver based TypeInfo can keep its PgTypeId undecided or whether it should follow mapping.DataTypeName.</param>
+delegate PgTypeInfo TypeInfoFactory(PgSerializerOptions options, TypeInfoMapping mapping, bool resolvedDataTypeName);
+
+readonly struct TypeInfoMapping
+{
+    public TypeInfoMapping(Type type, DataTypeName dataTypeName, bool isDefault, TypeInfoFactory factory)
+    {
+        Type = type;
+        DataTypeName = dataTypeName;
+        IsDefault = isDefault;
+        Factory = factory;
+    }
+
+    public TypeInfoFactory Factory { get; }
+    public Type Type { get; }
+    public DataTypeName DataTypeName { get; }
+    public bool IsDefault { get; }
+
+    public PgTypeInfo? GetConverterInfo(Type? type, DataTypeName? dataTypeName, PgSerializerOptions options)
+    {
+        var dataTypeMatch = dataTypeName == DataTypeName;
+        var typeMatch = type == Type;
+
+        if (dataTypeMatch && typeMatch)
+            return Factory(options, this, resolvedDataTypeName: true);
+
+        // We only match as a default if the respective data wasn't passed.
+        if (IsDefault && (dataTypeMatch && type is null || typeMatch && dataTypeName is null))
+            return Factory(options, this, resolvedDataTypeName: dataTypeMatch);
+
+        // Not a match
+        return null;
+    }
+}
+
+readonly struct ConverterInfoMappingCollection
+{
+    readonly List<TypeInfoMapping> _items = new();
+
+    public ConverterInfoMappingCollection()
+    {
+    }
+
+    public IReadOnlyCollection<TypeInfoMapping> Items => _items;
+
+    TypeInfoMapping? TryFindMapping(Type type, DataTypeName dataTypeName)
```

Comment:
It's not unheard of to use the existence or absence of null as the bool in find patterns. It would need another creative name for Find + throw again as well.

Open to suggestions.

Reply:
Yep, the way it is now LGTM.